RHOARDOC-1529 Files to describe building a RHOAR application on an image

huntley opened the pull request:
https://issues.jboss.org/browse/RHOARDOC-1529

Looking for feedback that this is on the right track. Don't spare the feedback. Thank you!

tradej commented:
ok to test

tradej commented on docs/topics/proc_selecting-the-jdk-builder-image-for-your-application.adoc:
```diff
@@ -0,0 +1,57 @@
++
+[source,xml,options="nowrap",subs="attributes+"]
+----
+<fabric8.generator.from>{name-2si-image-openjdk8-rhel7}</fabric8.generator.from>
```
A thought, not necessarily criticism: It may be worth considering putting this information in the reference module and reformat that as a list instead of a table.

huntley replied:
Not entirely happy with the modules but I'm not sure your suggestion helps (me) :-) So simple in DITA :-)

tradej replied:
Right now, the flow of the procedure is interrupted with what is essentially a reference table in step 2. I think it would be easier to follow to put that reference table in a separate module and put a requirement on having selected the value from the table in prerequisites.

huntley replied:
@tradej I don't wholly disagree but I feel that sending the reader to another module with no clear navigation back is not a great alternative. I made the second step less intrusive so let me know what you think.

tradej replied:
@huntley As Adam and Lada are suggesting, the content in this procedure makes sense as part of the _Preparing..._ module, so if you read _Deploying a Vert.x application to OpenShift_ from top to bottom, you read:

1. the assembly introduction (which can explain that you will need the URL for the POM file), then 
1. the table (that would also contain the URLs) in the _OpenJDK images for Red Hat..._ reference module,
1. in _Preparing...,_ you read a mention of the URL (that you're already aware of) in the prerequisites and step 2 only says to input the URL.

This way, the reader doesn't have to backtrack at all.

The independence of the individual modules is not violated because the _OpenJDK images for Red Hat..._ module can easily be replaced with any other module that contains URLs to input in the POM file.

An added benefit that I've just realised is that when you get another combination in the support matrix (a new version of JDK or some such), you only have to update one module.

In any case, I maintain that I don't think your solution is bad or that it must be changed, I just believe there's a more elegant way of going about it.

tradej commented on docs/topics/assembly_developing-and-deploying-vertx-application.adoc:
```diff
@@ -20,6 +20,12 @@ In addition to xref:example-rest-http-vertx[using an example], you can create ne
 //Developing an application
 include::proc_developing-runtime-application.adoc[leveloffset=+1]
 
+//Supported OS and JDK images
```
This doesn't fit very well here IMO. If this is for OpenShift only, I think it should be in the OpenShift deployment assembly..

huntley replied:
Moved.

tradej commented on docs/topics/assembly_developing-and-deploying-vertx-application.adoc:
```diff
@@ -20,6 +20,12 @@ In addition to xref:example-rest-http-vertx[using an example], you can create ne
+include::ref_openjdk-images-for-red-hat-enterprise-linux.adoc[leveloffset=+1]
+
+//Selecting the image to build and deploy your application
+include::proc_selecting-the-jdk-builder-image-for-your-application.adoc[leveloffset=+2]
```
If the `leveloffset=+2` is intentional, then this module and the previous one must be in a new assembly, both with `leveloffset=+1`. ("nested" modules are not allowed under FCC, all adjacent modules must be on the same level)

huntley replied:
Ok. Are the FCC "rules" published somewhere?

tradej replied:
https://redhat-documentation.github.io/modular-docs/

huntley replied:
@tradej I don't see that nested modules are forbidden. Perhaps that is more to do with FCC than modular docs.

tradej replied:
That seems to be an unfortunate omission then. It has been the idea from the beginning (assemblies being the only files allowed to include other files). I'll file a JIRA for an inclusion of such a note.

tradej replied:
Filed: redhat-documentation/modular-docs#99.

tradej commented on docs/topics/ref_openjdk-images-for-red-hat-enterprise-linux.adoc:
```diff
@@ -0,0 +1,26 @@
+[id='openjdk-images-for-red-hat-enterprise-linux_{context}']
+= OpenJDK images for Red Hat Enterprise Linux
+You can also use an OpenJDK image to build and deploy your application to OpenShift. Images are available for RHEL7 and RHEL 8 with OpenJDK 8 or OpenJDK 11.
```
"also" as opposed to what?

huntley replied:
Fixed.

tradej commented:
Mine are only random comments, please secure a full peer review, @huntley.

snelluli1 commented on docs/topics/proc_selecting-the-jdk-builder-image-for-your-application.adoc:
```diff
@@ -0,0 +1,57 @@
+[id='selecting-the-jdk-builder-image-for-your-application_{context}']
+= Selecting a JDK image for your {runtime} application
+Edit your application `pom.xml` file to specify the image for building and running your {runtime} application.
```
Edit your application `pom.xml` file--> Edit your application's `pom.xml` file?

tradej replied:
Edit the `pom.xml` file of/in your application. We're not supposed to use Saxon genitive with inanimate objects [IBM Style Guide].

huntley replied:
No Saxon's here.

tradej replied:
I was reacting to Sreelatha's comment. Still, I do believe _pom.xml file of/in your application_ is more correct here, or at least more consistent with the rest of the docs.

huntley replied:
@tradej. Understood and I revised based on your comment.

snelluli1 commented on docs/topics/proc_selecting-the-jdk-builder-image-for-your-application.adoc:
```diff
@@ -0,0 +1,57 @@
+Edit your application `pom.xml` file to specify the image for building and running your {runtime} application.
+
+.Prerequisites
+* A `pom.xml` file for your application.
```
This prerequisite is a little vague. Do we want to say that the user should have a pom.xml ready before they start this task?
Consider rephrasing to something like: The `pom.xml` file is created.

huntley replied:
Fixed. Thanks for the suggestion @inoxx03.

snelluli1 commented on docs/topics/ref_openjdk-images-for-red-hat-enterprise-linux.adoc:
```diff
@@ -0,0 +1,30 @@
+|link:https://access.redhat.com/containers/?tab=overview#/registry.access.redhat.com/openjdk/openjdk-11-rhel8[RHEL 8 with OpenJDK^]
+|===
+
+NOTE: The use of RHEL 8-based containers on a RHEL 7 host has limited support, which includes OpenShift 3 and OpenShift 4 installed on RHEL 7.
```
For better clarity, how about rephrasing this sentence as:
The use of RHEL 8-based containers on a RHEL 7 host, for example, OpenShift 3 and OpenShift 4 installed on RHEL 7, has limited support.
For more information, see the link:https://access.redhat.com/articles/4087871[RHEL Container Compatibility Matrix].

abrennan89 replied:
"OpenShift 3 or OpenShift 4 installed on a RHEL 7 host, is not fully supported." maybe? Just a suggestion :)

huntley replied:
Shorter is invariably better. As a colleague once told me, never use a short word when a diminutive one will do.

tradej replied:
Please check with PM about the support statements both in general and in context of the community version. Getting this right is imperative.

huntley replied:
@tradej Vert.x PM suggested the note. Will make sure he reviews the final version.

tradej commented on docs/topics/assembly_deploying-runtime-application-to-openshift.adoc:
```diff
@@ -6,6 +6,9 @@
 To deploy your {runtime} application to OpenShift, configure the `pom.xml` file in your application and then deploy using the Fabric8 Maven plugin.
 
+//Supported OS and JDK images
+include::ref_openjdk-images-for-red-hat-enterprise-linux.adoc[leveloffset=+1]
```
@inoxx03 I don't know if we're talking about the same thing. Asciidoctor resolves include statements relative to the file they're placed in, so I don't think this should cause any trouble with Thorntail as the assembly is in `/docs/topics` in either case.

huntley commented on docs/topics/proc_selecting-the-jdk-builder-image-for-your-application.adoc:
```diff
@@ -0,0 +1,57 @@
+
+. Replace the `fabric8.generator.from` property to specify the version of RHEL and OpenJDK you intend to use.
+
+RHEL 7 with OpenJKD 8::
```
Good catch!

Ladicek commented on docs/topics/ref_openjdk-images-for-red-hat-enterprise-linux.adoc:
```diff
@@ -0,0 +1,30 @@
+|link:https://access.redhat.com/containers/?tab=overview#/registry.access.redhat.com/openjdk/openjdk-11-rhel8[RHEL 8 with OpenJDK^]
+|===
+
+NOTE: The use of a RHEL{nbsp}8-based container on a RHEL{nbsp}7 host, for example OpenShift 3 or OpenShift 4, installed on RHEL{nbsp}7 has limited support.
```
Perhaps one little comment: this sentence "use of a RHEL 8-based container on a RHEL 7 host, for example OpenShift 3 or OpenShift 4, installed on RHEL 7 has limited support" needlessly mentions RHEL 7 twice.

huntley replied:
Totally agree.

akoniar commented on docs/topics/ref_openjdk-images-for-red-hat-enterprise-linux.adoc:
```diff
@@ -0,0 +1,30 @@
+|OpenJDK 11
+|link:https://access.redhat.com/containers/#/registry.access.redhat.com/openjdk/openjdk-11-rhel7[RHEL 7 with OpenJDK 11^]
+
+|RHEL 8
```
The rhel8 images are accessible only with authentication. Please make a note about it.

akoniar commented on docs/topics/proc_selecting-the-jdk-builder-image-for-your-application.adoc:
```diff
@@ -0,0 +1,57 @@
++
+[source,bash,options="nowrap",subs="attributes+"]
+----
+{parameter-maven-command}
```
I am not sure if `mvn {runtime}:run` starts a fabric8-maven so the last two steps are useless.
Also, it seems as we want to run the app as a standalone app.
Please could you move this to the section :
> Preparing {runtime} application for OpenShift deployment
WDYT @Ladicek

Ladicek replied:
Yea, this is all about OpenShift, doesn't make sense outside. I agree it would be best to merge this with the related content in the "OpenShift deployment" chapter.

huntley commented:
I addressed comments from @Ladicek and @akoniar (thank you!) and the peer reviews. Thank you all. Please provide feedback if I did not correctly interpret your comment, I will gladly make the change.

akoniar commented:
@huntley I still miss the mention of the Docker's (or podman) authentication into  registry.redhat.io for getting openjdk-${java.version}-rhel-8 images. I'd put it in the prerequisites. Otherwise LGTM

tradej commented on docs/topics/assembly_deploying-runtime-application-to-openshift.adoc:
```diff
@@ -4,7 +4,11 @@
 = Deploying {runtime} application to OpenShift
```
Anchor ID missing

huntley commented:
@inoxx03 @akoniar @Ladicek 

Sorry for the constant flogging but I would appreciate one final ACK on these changes. The most recent change centers on the need for a registry service account.

Thanks!

ncbaratta commented:
@huntley I'm still not seeing the prerequisite that @akoniar recommends in https://github.com/fabric8-launcher/launcher-documentation/pull/1291#issuecomment-499772562

huntley commented:
@akoniar Added prereq and modified sentence in assembly to make less confusing. Also updated table as I spotted an error there as well. Thanks for your help!

huntley commented:
@ncbaratta Chatted with @akoniar to resolve his outstanding comments. Thanks for the note, I missed that comment in combining the two procedures.

akoniar commented:
Changes are OK. Thank you @huntley
